Migration crash loop: LoginsStorageException: The logins tables are not empty

I accidentally updated Firefox Nightly on my phone to Fenix. Now I am in a migration crash loop. Whenever I start the app the migration starts and then after some seconds the app crashes.

```
         AndroidRuntime  E  FATAL EXCEPTION: main
                         E  Process: org.mozilla.fennec_aurora, PID: 13350
                         E  mozilla.appservices.logins.LoginsStorageException: The logins tables are not empty
                         E      at mozilla.appservices.logins.rust.RustError.intoException(RustError.kt:4)
                         E      at mozilla.appservices.logins.DatabaseLoginsStorage.importLogins(DatabaseLoginsStorage.kt:12)
                         E      at mozilla.components.service.sync.logins.AsyncLoginsStorageAdapter$importLoginsAsync$1.invokeSuspend(AsyncLoginsStorage
                            .kt:6)
                         E      at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:2)
                         E      at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:15)
                         E      at kotlinx.coroutines.scheduling.CoroutineScheduler.runSafely(CoroutineScheduler.kt:1)
                         E      at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.run(CoroutineScheduler.kt:15)
```

--- a/components/support/migration/src/main/java/mozilla/components/support/migration/FennecLoginsMigration.kt
+++ b/components/support/migration/src/main/java/mozilla/components/support/migration/FennecLoginsMigration.kt
@@ -10,6 +10,7 @@ import android.database.sqlite.SQLiteDatabase.OPEN_READONLY
 import android.util.Base64
 import androidx.annotation.VisibleForTesting
 import mozilla.components.lib.crash.CrashReporter
+import mozilla.components.service.sync.logins.AsyncLoginsStorage
 import mozilla.components.service.sync.logins.ServerPassword
 import mozilla.components.support.base.log.logger.Logger
 import mozilla.components.support.ktx.kotlin.pkcs7unpad
@@ -29,7 +30,11 @@ internal sealed class LoginsMigrationResult {
     internal sealed class Success : LoginsMigrationResult() {
         internal object MasterPasswordIsSet : Success()
 
-        internal data class LoginRecords(val records: List<ServerPassword>, val recordsDetected: Int) : Success()
+        internal data class ImportedLoginRecords(
+            val totalRecordsDetected: Long,
+            val failedToProcess: Long,
+            val failedToImport: Long
+        ) : Success()
     }
 
     internal sealed class Failure {
@@ -56,8 +61,28 @@ internal sealed class LoginsMigrationResult {
                 return "Unexpected metadata key material encryption alg oid: $berDecoded"
             }
         }
+
+        internal data class GetLoginsThrew(val e: Exception) : Failure() {
+            override fun toString(): String {
+                return "getLogins unexpectedly threw: $e"
+            }
+        }
+
+        internal data class RustImportThrew(val e: Exception) : Failure() {
+            override fun toString(): String {
+                return "Rust import unexpectedly threw: $e"
+            }
+        }
     }
 }
+
+@VisibleForTesting
+internal data class FennecLoginRecords(
+    // Records that we were able to process.
+    val records: List<ServerPassword>,
+    // Total number of records that we saw in the database.
+    val totalRecordsDetected: Int
+)
 
 @Suppress("TooManyFunctions", "LargeClass")
 internal object FennecLoginsMigration {
@@ -67,11 +92,13 @@ internal object FennecLoginsMigration {
     private val logger = Logger("FennecLoginsMigration")
     private val keyForId: MutableMap<String, ByteArray> = mutableMapOf()
 
-    @Suppress("TooGenericExceptionCaught", "ReturnCount")
-    internal fun migrate(
+    @Suppress("TooGenericExceptionCaught", "ReturnCount", "LongParameterList")
+    internal suspend fun migrate(
         crashReporter: CrashReporter,
         signonsDbPath: String,
         key4DbPath: String,
+        loginsStorage: AsyncLoginsStorage,
+        loginsStorageKey: String,
         masterPassword: String = DEFAULT_MASTER_PASSWORD
     ): Result<LoginsMigrationResult> {
         val noMasterPassword = try {
@@ -85,13 +112,26 @@ internal object FennecLoginsMigration {
             return Result.Success(LoginsMigrationResult.Success.MasterPasswordIsSet)
         }
 
-        val loginRecords = try {
+        val fennecRecords = try {
             getLogins(crashReporter, masterPassword, signonsDbPath = signonsDbPath, key4DbPath = key4DbPath)
         } catch (e: Exception) {
-            return Result.Failure(e)
+            return Result.Failure(LoginMigrationException(LoginsMigrationResult.Failure.GetLoginsThrew(e)))
         }
 
-        return Result.Success(loginRecords)
+        loginsStorage.ensureUnlocked(loginsStorageKey).await()
+        val failedToImport = try {
+            loginsStorage.importLoginsAsync(fennecRecords.records).await()
+        } catch (e: Exception) {
+            return Result.Failure(LoginMigrationException(LoginsMigrationResult.Failure.RustImportThrew(e)))
+        } finally {
+            loginsStorage.ensureLocked().await()
+        }
+
+        return Result.Success(LoginsMigrationResult.Success.ImportedLoginRecords(
+            totalRecordsDetected = fennecRecords.totalRecordsDetected.toLong(),
+            failedToProcess = (fennecRecords.totalRecordsDetected - fennecRecords.records.size).toLong(),
+            failedToImport = failedToImport
+        ))
     }
 
     /**
@@ -123,12 +163,13 @@ internal object FennecLoginsMigration {
     }
 
     @Suppress("LongMethod")
-    private fun getLogins(
+    @VisibleForTesting
+    internal fun getLogins(
         crashReporter: CrashReporter,
         masterPassword: String,
         signonsDbPath: String,
         key4DbPath: String
-    ): LoginsMigrationResult.Success.LoginRecords {
+    ): FennecLoginRecords {
         val db = SQLiteDatabase.openDatabase(signonsDbPath, null, OPEN_READONLY)
         val records: MutableList<ServerPassword> = mutableListOf()
 
@@ -216,7 +257,7 @@ internal object FennecLoginsMigration {
             crashReporter.submitCaughtException(it.value)
         }
 
-        return LoginsMigrationResult.Success.LoginRecords(records = records, recordsDetected = totalRecordCount)
+        return FennecLoginRecords(records = records, totalRecordsDetected = totalRecordCount)
     }
 
     private fun SignonsEntry.decrypt(masterPassword: String, key4DbPath: String): ByteArray {

--- a/components/support/migration/src/main/java/mozilla/components/support/migration/FennecMigrator.kt
+++ b/components/support/migration/src/main/java/mozilla/components/support/migration/FennecMigrator.kt
@@ -501,7 +501,9 @@ class FennecMigrator private constructor(
             FennecLoginsMigration.migrate(
                 crashReporter,
                 signonsDbPath = "${profile.path}/$signonsDbName",
-                key4DbPath = "${profile.path}/$key4DbName"
+                key4DbPath = "${profile.path}/$key4DbName",
+                loginsStorage = loginsStorage!!,
+                loginsStorageKey = loginsStorageKey!!
             )
         } catch (e: Exception) {
             crashReporter.submitCaughtException(FennecMigratorException.MigrateLoginsException(e))
@@ -530,6 +532,16 @@ class FennecMigrator private constructor(
                     crashReporter.submitCaughtException(migrationFailureWrapper)
                     result
                 }
+                is LoginsMigrationResult.Failure.GetLoginsThrew -> {
+                    logger.error("getLogins failure: $failure")
+                    crashReporter.submitCaughtException(migrationFailureWrapper)
+                    result
+                }
+                is LoginsMigrationResult.Failure.RustImportThrew -> {
+                    logger.error("Rust import failure: $failure")
+                    crashReporter.submitCaughtException(migrationFailureWrapper)
+                    result
+                }
             }
         }
 
@@ -540,22 +552,12 @@ class FennecMigrator private constructor(
                 result
             }
 
-            is LoginsMigrationResult.Success.LoginRecords -> {
-                logger.debug(
-                    "Parsed ${success.records.size} out of ${success.recordsDetected} login records. Importing..."
-                )
-
-                loginsStorage!!.ensureUnlocked(loginsStorageKey!!).await()
-                try {
-                    val failedToImport = loginsStorage.importLoginsAsync(success.records).await()
-                    if (failedToImport > 0L) {
-                        logger.debug("Failed to import $failedToImport")
-                    } else {
-                        logger.debug("Imported all records")
-                    }
-                } finally {
-                    loginsStorage.ensureLocked().await()
-                }
+            is LoginsMigrationResult.Success.ImportedLoginRecords -> {
+                logger.debug("""Imported login records! Details:
+                    Total detected=${success.totalRecordsDetected},
+                    failed to process=${success.failedToProcess},
+                    failed to import=${success.failedToImport}
+                """.trimIndent())
                 result
             }
         }

--- a/components/support/migration/src/test/java/mozilla/components/support/migration/FennecLoginsMigrationTest.kt
+++ b/components/support/migration/src/test/java/mozilla/components/support/migration/FennecLoginsMigrationTest.kt
@@ -53,14 +53,13 @@ class FennecLoginsMigrationTest {
     @Test
     fun `decrypted records without a set master password`() {
         val crashReporter: CrashReporter = mock()
-        with(FennecLoginsMigration.migrate(
+        with(FennecLoginsMigration.getLogins(
             crashReporter,
+            masterPassword = FennecLoginsMigration.DEFAULT_MASTER_PASSWORD,
             signonsDbPath = File(getTestPath("logins"), "basic/signons.sqlite").absolutePath,
             key4DbPath = File(getTestPath("logins"), "basic/key4.db").absolutePath
-        ) as Result.Success) {
-            assertEquals(LoginsMigrationResult.Success.LoginRecords::class, this.value::class)
-            val successResult = (this.value as LoginsMigrationResult.Success.LoginRecords)
-            assertEquals(2, successResult.recordsDetected)
+        )) {
+            assertEquals(2, this.totalRecordsDetected)
             assertEquals(listOf(
                 ServerPassword(
                     id = "{cf7cabe4-ec82-4800-b077-c6d97ffcd63a}",
@@ -90,7 +89,7 @@ class FennecLoginsMigrationTest {
                     usernameField = "",
                     passwordField = ""
                 )
-            ), successResult.records)
+            ), this.records)
         }
         verifyZeroInteractions(crashReporter)
     }
@@ -98,14 +97,13 @@ class FennecLoginsMigrationTest {
     @Test
     fun `decrypted record with long encoded strings without a set master password`() {
         val crashReporter: CrashReporter = mock()
-        with(FennecLoginsMigration.migrate(
+        with(FennecLoginsMigration.getLogins(
             crashReporter,
+            masterPassword = FennecLoginsMigration.DEFAULT_MASTER_PASSWORD,
             signonsDbPath = File(getTestPath("logins"), "basic/longPass-signons.sqlite").absolutePath,
             key4DbPath = File(getTestPath("logins"), "basic/longPass-key4.db").absolutePath
-        ) as Result.Success) {
-            assertEquals(LoginsMigrationResult.Success.LoginRecords::class, this.value::class)
-            val successResult = (this.value as LoginsMigrationResult.Success.LoginRecords)
-            assertEquals(2, successResult.recordsDetected)
+        )) {
+            assertEquals(2, this.totalRecordsDetected)
             assertEquals(listOf(
                 ServerPassword(
                     id = "{0a5e8c6c-11bd-49e4-a09f-cf33c3b65498}",
@@ -138,7 +136,7 @@ class FennecLoginsMigrationTest {
                     usernameField = "",
                     passwordField = ""
                 )
-            ), successResult.records)
+            ), this.records)
         }
         verifyZeroInteractions(crashReporter)
     }
@@ -146,15 +144,13 @@ class FennecLoginsMigrationTest {
     @Test
     fun `decrypted records with a master password`() {
         val crashReporter: CrashReporter = mock()
-        with(FennecLoginsMigration.migrate(
+        with(FennecLoginsMigration.getLogins(
             crashReporter,
+            masterPassword = "my secret pass",
             signonsDbPath = File(getTestPath("logins"), "with-mp/signons.sqlite").absolutePath,
-            key4DbPath = File(getTestPath("logins"), "with-mp/key4.db").absolutePath,
-            masterPassword = "my secret pass"
-        ) as Result.Success) {
-            assertEquals(LoginsMigrationResult.Success.LoginRecords::class, this.value::class)
-            val successResult = (this.value as LoginsMigrationResult.Success.LoginRecords)
-            assertEquals(2, successResult.recordsDetected)
+            key4DbPath = File(getTestPath("logins"), "with-mp/key4.db").absolutePath
+        )) {
+            assertEquals(2, this.totalRecordsDetected)
             assertEquals(listOf(
                 ServerPassword(
                     id = "{cf7cabe4-ec82-4800-b077-c6d97ffcd63a}",
@@ -184,7 +180,7 @@ class FennecLoginsMigrationTest {
                     usernameField = "",
                     passwordField = ""
                 )
-            ), successResult.records)
+            ), this.records)
         }
         verifyZeroInteractions(crashReporter)
     }

